### PR TITLE
MINOR: Add test for PartitionMetadataFile

### PR DESCRIPTION
**Pull request by KevinZTW**
In PartitionMetadataFile.java, there is a defensive early-fail design when setting a different topicId, but there are no corresponding test cases. Hence, this PR adds two tests for setting the same and different topicId.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: storage/src/test/java/org/apache/kafka/storage/internals/checkpoint/PartitionMetadataFileTest.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+import org.junit.jupiter.api.Test;
+
+import java.io.File;
+import java.nio.file.Files;
+
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+public class PartitionMetadataFileTest  {
+    private final File dir = assertDoesNotThrow(() -> Files.createTempDirectory("tmp")).toFile();
```
Comment (chia7712) [inline storage/src/test/java/org/apache/kafka/storage/internals/checkpoint/PartitionMetadataFileTest.java]: we can replace it by `TestUtils.tempDirectory()`, and it can handle the cleanup of temporary folder we created.

**Review by chia7712 — COMMENTED** [review]
@KevinZTW thanks for enhancing the test coverage. Could you also add unit tests for other methods? thanks

File: storage/src/test/java/org/apache/kafka/storage/internals/checkpoint/PartitionMetadataFileTest.java (hunk 1 above)
Comment (brandboat) [reply storage/src/test/java/org/apache/kafka/storage/internals/checkpoint/PartitionMetadataFileTest.java]: looks like we have to add this `<allow pkg="org.apache.kafka.test" />` to 
https://github.com/apache/kafka/blob/c034cf2953e691ce4ecd94bf00ac5810167354bc/checkstyle/import-control-storage.xml#L73-L80

otherwise checkstyle raise error `disallow import`

Comment (KevinZTW) [reply storage/src/test/java/org/apache/kafka/storage/internals/checkpoint/PartitionMetadataFileTest.java]: @chia7712 Thanks! I did try with that one and received the error message, thought that is intentionally forbidden.
@brandboat Much thanks for pointing me the way!

File: storage/src/test/java/org/apache/kafka/storage/internals/checkpoint/PartitionMetadataFileTest.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    public void testMaybeFlushWithTopicIdPresent() {
+        File file = PartitionMetadataFile.newFile(dir);
+        PartitionMetadataFile partitionMetadataFile = new PartitionMetadataFile(file, null);
+
+        Uuid topicId = Uuid.randomUuid();
+        assertDoesNotThrow(() -> partitionMetadataFile.record(topicId));
+        assertDoesNotThrow(partitionMetadataFile::maybeFlush);
+
+        assertDoesNotThrow(() -> {
+            List<String> lines = Files.readAllLines(file.toPath());
```
Comment (KevinZTW) [inline storage/src/test/java/org/apache/kafka/storage/internals/checkpoint/PartitionMetadataFileTest.java]: Since the temp file would be created and swap with metadata file, I found it a bit hard to directly verify related behavior. So I test on the end result instead, hope that is the right approach.

**Comment by KevinZTW** [comment]
>@KevinZTW thanks for enhancing the test coverage. Could you also add unit tests for other methods? thanks

@chia7712 no problem! I have added test cases to cover other methods, could you help me to take another look?

File: storage/src/test/java/org/apache/kafka/storage/internals/checkpoint/PartitionMetadataFileTest.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+import java.io.File;
+import java.nio.file.Files;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+public class PartitionMetadataFileTest  {
+    private final File dir = TestUtils.tempDirectory();
```
Comment (chia7712) [inline storage/src/test/java/org/apache/kafka/storage/internals/checkpoint/PartitionMetadataFileTest.java:38]: It seems `File file = PartitionMetadataFile.newFile(dir)` is more suitable as it is used by all test cases.

File: storage/src/test/java/org/apache/kafka/storage/internals/checkpoint/PartitionMetadataFileTest.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        File file = PartitionMetadataFile.newFile(dir);
+        PartitionMetadataFile partitionMetadataFile = new PartitionMetadataFile(file, null);
+
+        Uuid topicId = Uuid.randomUuid();
+        assertDoesNotThrow(() -> partitionMetadataFile.record(topicId));
+        assertDoesNotThrow(partitionMetadataFile::maybeFlush);
+
+        assertDoesNotThrow(() -> {
+            List<String> lines = Files.readAllLines(file.toPath());
+            assertEquals(2, lines.size());
```
Comment (chia7712) [inline storage/src/test/java/org/apache/kafka/storage/internals/checkpoint/PartitionMetadataFileTest.java]: it is a bit weird that we call `assert` in the `assertDoesNotThrow`. It should be fine to expose the exception to method signature.

File: storage/src/test/java/org/apache/kafka/storage/internals/checkpoint/PartitionMetadataFileTest.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+            assertEquals("version: 0", lines.get(0));
```
Comment (chia7712) [inline storage/src/test/java/org/apache/kafka/storage/internals/checkpoint/PartitionMetadataFileTest.java]: it would be nice to add comments about those contents are from `PartitionMetadata#encode`. otherwise, it would be hard to debug if this test is broken by something changes in the future.

File: storage/src/test/java/org/apache/kafka/storage/internals/checkpoint/PartitionMetadataFileTest.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    @Test
+    public void testRead() {
+        File file = PartitionMetadataFile.newFile(dir);
+        LogDirFailureChannel channel = Mockito.mock(LogDirFailureChannel.class);
+        PartitionMetadataFile partitionMetadataFile = new PartitionMetadataFile(file, channel);
+
+        Uuid topicId = Uuid.randomUuid();
+        assertDoesNotThrow(() -> partitionMetadataFile.record(topicId));
```
Comment (chia7712) [inline storage/src/test/java/org/apache/kafka/storage/internals/checkpoint/PartitionMetadataFileTest.java]: we don't need to use `assertDoesNotThrow` since it is fine to add any exception to method signature.

**Review by chia7712 — COMMENTED** [review]
@KevinZTW thanks for updated PR. a couple of comments left.

File: storage/src/test/java/org/apache/kafka/storage/internals/checkpoint/PartitionMetadataFileTest.java (hunk 4 above)
Comment (KevinZTW) [reply storage/src/test/java/org/apache/kafka/storage/internals/checkpoint/PartitionMetadataFileTest.java]: I see, thanks!

File: storage/src/test/java/org/apache/kafka/storage/internals/checkpoint/PartitionMetadataFileTest.java (hunk 5 above)
Comment (KevinZTW) [reply storage/src/test/java/org/apache/kafka/storage/internals/checkpoint/PartitionMetadataFileTest.java]: ok! thanks

File: storage/src/test/java/org/apache/kafka/storage/internals/checkpoint/PartitionMetadataFileTest.java (hunk 6 above)
Comment (KevinZTW) [reply storage/src/test/java/org/apache/kafka/storage/internals/checkpoint/PartitionMetadataFileTest.java]: Oh, in here the reason I used `assertDoesNotThrow` is because we expect the first time to `record` the `topicId` shouldn't generate any exception, since there's no `dirtyTopicIdOpt`

Do you think that make sense? Or that would be redundant?

Comment (KevinZTW) [reply storage/src/test/java/org/apache/kafka/storage/internals/checkpoint/PartitionMetadataFileTest.java]: The only exception would be generated is `InconsistentTopicIdException` in this case I think

File: storage/src/test/java/org/apache/kafka/storage/internals/checkpoint/PartitionMetadataFileTest.java (hunk 3 above)
Comment (KevinZTW) [reply storage/src/test/java/org/apache/kafka/storage/internals/checkpoint/PartitionMetadataFileTest.java:38]: Oh thanks! I forgot the hook would cleanup files in each test automatically

File: storage/src/test/java/org/apache/kafka/storage/internals/checkpoint/PartitionMetadataFileTest.java (hunk 6 above)
Comment (chia7712) [reply storage/src/test/java/org/apache/kafka/storage/internals/checkpoint/PartitionMetadataFileTest.java]: >I used assertDoesNotThrow is because we expect the first time to record the topicId shouldn't generate any exception, since there's no dirtyTopicIdOpt

I agree to that. My point was `assertDoesNotThrow` converts the checked exception to unchecked exception, and it is useful when we run assert in the lambda. However, in those test case we can add the exception to method signature, and just let it fail. 

In short, `assertDoesNotThrow` in those test cases is redundant wrapper.

Comment (KevinZTW) [reply storage/src/test/java/org/apache/kafka/storage/internals/checkpoint/PartitionMetadataFileTest.java]: I see, thank you very much for the detailed explanation! I've revised it as suggested.

**Review by chia7712 — APPROVED** [review]
LGTM

File: storage/src/test/java/org/apache/kafka/storage/internals/checkpoint/PartitionMetadataFileTest.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+        Uuid differentTopicId = Uuid.randomUuid();
+        assertThrows(InconsistentTopicIdException.class, () -> partitionMetadataFile.record(differentTopicId));
+    }
+
+    @Test
+    public void testSetRecordWithSameTopicId() {
+        PartitionMetadataFile partitionMetadataFile = new PartitionMetadataFile(file, null);
+        Uuid topicId = Uuid.randomUuid();
+        partitionMetadataFile.record(topicId);
+        assertDoesNotThrow(() -> partitionMetadataFile.record(topicId));
```
Comment (chia7712) [inline storage/src/test/java/org/apache/kafka/storage/internals/checkpoint/PartitionMetadataFileTest.java:55]: we need test to call record with UUID having different reference and same content.

**Review by chia7712 — COMMENTED** [review]
@KevinZTW thanks for updated PR. please take a look comment to add more tests

File: storage/src/test/java/org/apache/kafka/storage/internals/checkpoint/PartitionMetadataFileTest.java (hunk 7 above)
Comment (KevinZTW) [reply storage/src/test/java/org/apache/kafka/storage/internals/checkpoint/PartitionMetadataFileTest.java:55]: sorry I missed this one.... just update the PR as suggested!

**Review by chia7712 — APPROVED** [review]
will merge it after QA pass
